### PR TITLE
v4

### PR DESCRIPTION
**Pull request by goosewobbler**
- [x] Update deps & WDIO 
- [x] Finish ESM / CJS approach
- [x] Auto-download Chromedriver
- [x] Move example package into repo => integration tests
- [x] Improve API interface wrt types
- [x] Jest => Vitest
- [x] Fix existing unit tests
- [x] Fix existing unit tests on CI
- [x] Integration tests running on CI
- [x] CI on all platforms
- [x] Fix Integration tests running on CI
- [x] Fix Windows CI build
- [x] Add workaround for #50 
- [x] Fix pre-push hook
- [x] Update ESLint config, fix issues
- [x] Update docs (chromedriver handling)
- [x] Update docs (API)
- [x] Re-enable images in chromeArgs
- [x] Add error case & test for no chromedriverCustomPath or electronVersion set
- [x] Add unit tests for Chromedriver auto-download
- [x] Add unit tests for Chromedriver download failure
- [x] Fix CI test timeout
- [x] Test on Intel Mac

### CODE REVIEW

File: scripts/build.sh [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+cat >./dist/package.json <<!EOF
+{
+    "type": "module"
+}
+!EOF
+
+cat >./dist/cjs/package.json <<!EOF
+{
+    "type": "commonjs"
+}
```
Comment (christian-bromann) [inline scripts/build.sh:12]: Another option would be to make the project an ESM project and have a `/src/cjs` folder with a `package.json` that defines the `cjs` folder as CJS project like we have in the [chromedriver service](https://github.com/webdriverio-community/wdio-chromedriver-service/tree/main/src/cjs).

File: example/tsconfig.base.json [hunk 2]
```diff
@@ -0,0 +1,25 @@
+{
+  "compilerOptions": {
```
Comment (christian-bromann) [inline example/tsconfig.base.json:2]: can a bunch of this stuff be inherited from the base tsconfig in the root directory?

File: src/index.ts [hunk 3]
```diff
@@ -1,5 +1,11 @@
...
+import ElectronWorkerService from './service.js';
 
 export default ElectronWorkerService;
 export const launcher = ChromedriverLauncher;
+export type electronService = {
+  api: (...arg: unknown[]) => Promise<unknown>;
+  app: (funcName: string, ...arg: unknown[]) => Promise<unknown>;
+  mainProcess: (funcName: string, ...arg: unknown[]) => Promise<unknown>;
+  browserWindow: (funcName: string, ...arg: unknown[]) => Promise<unknown>;
+} & WebdriverIO.Browser;
```
Comment (christian-bromann) [inline src/index.ts:11]: It seems like we should extend the `WebdriverIO.Browser` interface to add these to it, e.g. like we do with the [`@wdio/shared-store-service`](https://github.com/webdriverio/webdriverio/blob/main/packages/wdio-shared-store-service/src/index.ts#L10-L22). Wdyt?

**Review by christian-bromann — APPROVED** [review]
Some comments but overall, great work!

File: src/index.ts (hunk 3 above)
Comment (goosewobbler) [reply src/index.ts]: I tried something like this, also went for a re-export of browser from `@wdio/globals` as I read wdio is moving away from globals.  Works well I think

File: scripts/build.sh (hunk 1 above)
Comment (goosewobbler) [reply scripts/build.sh:12]: I tried this approach first and couldn't get it to work with the example app approach I am using.  Given I have three package jsons in the repo now I'm still considering a monorepo-type approach using pnpm workspaces might be better if only to streamline deps & scripts.  The current package and ESM approach works for now so I am inclined to reduce scope and revisit later.

Comment (goosewobbler) [reply scripts/build.sh]: I found a way round the issues I was experiencing, the ESM approach is now close to that of the Chromedriver service.

**Comment by m4hdyar** [comment]
Thanks for your efforts. 
In service.ts there is a 
`const chromeArgs = [];

    if (isCI) {
      chromeArgs.push('window-size=1280,800');
      chromeArgs.push('blink-settings=imagesEnabled=false');
      chromeArgs.push('enable-automation');
      chromeArgs.push('disable-infobars');
      chromeArgs.push('disable-extensions');

      if (process.platform !== 'win32') {
        chromeArgs.push('no-sandbox');
        chromeArgs.push('disable-gpu');
        chromeArgs.push('disable-dev-shm-usage');
        chromeArgs.push('disable-setuid-sandbox');
      }
    }`

I understand that it's a workaround for  "DevToolsActivePort file doesn't exist". But in my case for example we don't have any issue in Gitlab but the line `chromeArgs.push('blink-settings=imagesEnabled=false');` makes images disabled. But we prefer that images remain enabled. Because of screenshots in reports. With the way it currently works, there isn't any option to edit these arguments. So do you think that we can also make this thing optional in V4 ?

**Comment by goosewobbler** [comment]
@m4hdyar You can probably override this in the current version by specifying `'blink-settings=imagesEnabled=true'` in `appArgs` but all the integration tests pass without so I think I will just delete this option and let people specify it in `appArgs` if required.

**Comment by m4hdyar** [comment]
> @m4hdyar You can probably override this in the current version by specifying `'blink-settings=imagesEnabled=true'` in `appArgs`. I just removed it on this branch and the integration tests pass without so I think I will just delete this option and let people specify it in `appArgs` if required.

Thanks 👍 I also think letting people decide about these options are better, even animation or window size. Because there isn't any possibility to override them.
I've already tried your suggestion but it didn't work. So I'm currently deleting CI environment variables in onPrepare to bypass ci-info as a workaround.

**Comment by christian-bromann** [comment]
Congrats on this excellent piece of work 👏 Amplified it on Twitter: https://twitter.com/webdriverio/status/1627087065283440641

**Pull request merged**